Rust rewrite

svmnotn opened the pull request:
No more runtimes

svmnotn committed:
start Rust rewrite

svmnotn committed:
add dependencies needed for the drivers

svmnotn committed:
add Operation enum and Motion struct to main.rs for later use

svmnotn committed:
add Driver trait and some test drivers

* created TestDriver with a test implementation
* created driver_postgres.rs to hold the postgres driver and implemented it with a test impl

svmnotn committed:
add Accelerator

still contains debug code, is it pretty enough? Those lines where only kept to see if things break.

svmnotn committed:
add test code to make sure that the accelerator is working properly

This is my version of tests... I have not learned those fancy shamncy tests that other people do with test functions and all that stuff.

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+    }
+    // Needs to be un pub.ed
+    pub fn within_bounds(&self, n: isize) -> isize{
+        println!("called bounds with \'{}\'", n);
```
Is this the best way to debug/throw errors?

svmnotn replied:
No.. just the best way I currently know + I needed something so that I could see it was doing things. Will be removed.

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+    pub fn new(mots: Vec<Motion>, drv: D) -> Self {
+        Accelerator{motions: mots, driver: drv}
+    }
+    // Needs to be un pub.ed
```
What does this mean? Why “need” vs “is”?

svmnotn replied:
because it is only marked as pub for testing reasons, it shouldn't be. pub means that it is public.

calebmer replied:
O, I was thinking Spanish with “un” and some file with “pub.ed” 😉

svmnotn replied:
... lol

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+    }
+
+    fn execute(&mut self, mut start: isize, mut finish: isize){
+        println!("called execute with \'{0}\' and \'{1}\'", start,finish);
```
Again, is this really the best way to debug?

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+
+    fn execute(&mut self, mut start: isize, mut finish: isize){
+        println!("called execute with \'{0}\' and \'{1}\'", start,finish);
+        if start != finish {
```
I generally like the pattern of `if (start === finish) { return }` as it reduces cyclomatic complexity. Is there a reason to do it this way?

svmnotn replied:
I just like the while !done pattern better...

calebmer replied:
? We’re talking about if-statements

svmnotn replied:
... ok, same thing _if_ !done

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+                    let mut i = start;
+                    loop{
+                        if i == finish { break }
+                        i += op;
```
That's a clever way to eliminate the need for `shift` 😊

Although it isn't too DRY. I'm not concerned, should I be?

svmnotn replied:
yes because I hate it.. but the shift wouldn't produce the correct result thus this was the best fix I could come up with... I'm still trying to think of a way of not having it be this way.

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+
+    pub fn add(&mut self){
+        println!("called add");
+        self.go(1);
```
This is only going up one. It should go to infinity.

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+
+    pub fn sub(&mut self){
+        println!("called sub");
+        self.go(-1);
```
This is only going down one. It should go to negative infinity.

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,98 @@
+        println!("called redo");
+        self.sub();
+        self.add();
+    }
```
`add()` & `sub()` should be going to infinity so this would need to be `go(-1)` & `go(1)`

calebmer commented on src/drivers/driver_postgres.rs:
```diff
@@ -0,0 +1,25 @@
+    target: &'static str,
+}
+
+impl Driv for Driver {
```
Difference between `Driv` and `Driver` is hard to understand. Maybe `Driver` and `PostgresDriver`?

svmnotn replied:
`use drivers::Driver as Driv;` that is the line that makes it understandable...
I simply didn't want the path to the postgres driver to be drivers::driver_postgres::PostgresDriver since it would take mean spelling postgres twice and I tend to spell it with more than one 's' from time to time. I could change it if it is a problem... I could also change the as to something like `DriverTrait`

calebmer replied:
`DriverTrait` would be better. Personally, I prefer verbosity whenever possible. If I have to choose between verbose variable names and comments everywhere, I'd choose verbose variables.

calebmer commented:
Those are some nice commit messages 😊

Ok, did some code reviewings, the method for debugging, `println!`, is that temporary? Also, where are the tests?

svmnotn commented:
The main method bit are the tests ;)
if you read the whole of the commit msg you would know.

calebmer commented:
I know, I just really don't like it 😊

I need formal tests to merge so we can determine when backwards compatibility is broken and set up a CI service.

svmnotn commented:
I'll do them when I learn how...

svmnotn committed:
rename driver_postgres to postgres

also I changed it to be DriverTrait for readability

svmnotn committed:
remove test code

svmnotn committed:
un publish within_bounds

also fix the space between start and finish in line 53

svmnotn committed:
move Motion and Operation into Accelerator

svmnotn committed:
add get_status to the accelerator

needed to see if the tests are doing the right thing

svmnotn committed:
add tests module

this module should mirror the outside of it so that the tests are organized per thing they test

svmnotn committed:
remove TestDriver from drivers/mod.rs

not needed since the tests mod implements it locally

svmnotn commented:
learned how to do tests in Rust :)
for informational purposes the command is `cargo test`

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,107 @@
+
+impl Operation{
+    fn get(finish: isize, start: isize) -> Self{
+        if finish < start { return Operation::Sub(-1) }
```
`-1` and `1` are constants in this case, is there anyway to code them that way?

svmnotn replied:
not that I know of

calebmer commented on src/main.rs:
```diff
@@ -0,0 +1,8 @@
+mod tests;
+
+fn main() {
+    println!("Hello World!");
```
Hello main function

svmnotn committed:
move Motion out into its own file

svmnotn committed:
rename go to shift and removed add and sub

Changed tests to reflect it

svmnotn committed:
change some tests

svmnotn committed:
allow for chaining of method calls

Learned that this is how one Rusts

svmnotn committed:
add clap as a dependency

clap is the closest rust lib I could find that does what commander does

svmnotn committed:
update dependencies

svmnotn committed:
stop my attempt to do OOP for this

svmnotn committed:
use String instead of the other kind...

Rust is rather confusing in all its string stuff

svmnotn committed:
add a default driver for when no compile objective is setup

This should never happen and might be removed

svmnotn committed:
update postgres driver and add reminder to populate it

svmnotn committed:
update accelerator tests to match none OOP accelerator

svmnotn committed:
update the Motion struct to match what is needed

svmnotn committed:
begin CLI work

svmnotn committed:
clamp makes a better name than within_bounds

svmnotn committed:
use DefaultDriver as TestDriver

svmnotn committed:
add reminder to implement

svmnotn committed:
change ls signature, remove redo, add create with reminder to implement it

svmnotn committed:
make the subcommands useable, however they lack a proper driver

svmnotn committed:
format the code

svmnotn committed:
begin implementation of motions::get

svmnotn committed:
fix spelling + change url to URL

svmnotn committed:
rename Motion::get to Motion::get_test

svmnotn committed:
clean imports

svmnotn committed:
unify error syntax/handling in main

Also added proper errors

svmnotn committed:
change ls to take a Vec<Motion>

also formatting and added myself as dev

svmnotn committed:
start dev on the IO side of motions

svmnotn committed:
use 'better' rust syntax

too many valid ways to write things ... still getting used to it all...

calebmer commented:
What's the progress? Do we need to talk?

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,89 @@
+    n
+}
+
+fn execute<D: Driver>(driver: &mut D, motions: &Vec<Motion>, mut start: isize, mut finish: isize) {
```
Should this be private to denote it should not be used outside of this module/file? Or at least marked as private with an underscore (`_execute`).

svmnotn replied:
Rust all is private by default, since it is not annotated as `pub` it won't be public

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,89 @@
+}
+
+impl Operation{
+    fn get(finish: isize, start: isize) -> Self {
```
Might this be better as its own function? So `get_operation`? Or does it make more sense as an operation?

svmnotn replied:
What do you mean?

I prefer the style of `Opetation::get()` rather than just `get_operation()`

calebmer commented on src/main.rs:
```diff
@@ -0,0 +1,111 @@
+                      .get_matches();
+
+    let target = matches.value_of("target").unwrap();
+    let directory = matches.value_of("directory").unwrap_or(".");
```
Wow, just wanted to say how javascripty this line looks 😉

calebmer commented on src/tests/accelerator.rs:
```diff
@@ -0,0 +1,147 @@
+    let mut drv = get_driver();
+    accelerator::reset(&mut drv, &get_motions());
+    assert_eq!(0, drv.get_status());
+}
```
Yay these are good looking tests. But Github is complaining there's no newline at file end?

svmnotn replied:
I like it that way.

Changed for GitHub and your preference...

svmnotn committed:
add rustfmt.toml

  this will override the default settings for rustfmt so that it uses longer lines

svmnotn committed:
update dependencies

svmnotn committed:
format the code base

svmnotn committed:
use clap macros

svmnotn committed:
begin implementation of motions

svmnotn committed:
format

svmnotn committed:
make rustfmt not touch strings

svmnotn committed:
make more explicit that the current functions in motions are test only

svmnotn committed:
use clap better

svmnotn committed:
use value_t_or_exit for better looks

Also do something when no sub command is passed

svmnotn committed:
formatting

svmnotn committed:
turn generics into Box<T> because generics do static dispatch

svmnotn committed:
compiles!

calebmer commented on src/accelerator.rs:
```diff
@@ -0,0 +1,87 @@
+  if n > max {
+    return max;
+  }
+  n
```
What is that?

svmnotn replied:
clamp? it clamps....

calebmer replied:
No the `n`
On Sun, Dec 20, 2015 at 9:25 AM Victor M. Suarez notifications@github.com
wrote:

> In src/accelerator.rs
> https://github.com/calebmer/accelerate/pull/2#discussion_r48103873:
> 
> > @@ -0,0 +1,87 @@
> > +use drivers::Driver;
> > +use motions::Motion;
> > +
> > +fn clamp(n: isize, min: isize, max: isize) -> isize {
> > -  if n < min {
> > -    return min;
> > -  }
> > -  if n > max {
> > -    return max;
> > -  }
> > -  n
> 
> clamp? it clamps....
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/calebmer/accelerate/pull/2/files#r48103873.

svmnotn replied:
n is the input, in rust you dont have to `return n;` you can just leave it at `n` and it means the same

calebmer replied:
I noticed, it doesn't have a semicolon 😊
On Sun, Dec 20, 2015 at 11:08 AM Victor M. Suarez notifications@github.com
wrote:

> In src/accelerator.rs
> https://github.com/calebmer/accelerate/pull/2#discussion_r48104873:
> 
> > @@ -0,0 +1,87 @@
> > +use drivers::Driver;
> > +use motions::Motion;
> > +
> > +fn clamp(n: isize, min: isize, max: isize) -> isize {
> > -  if n < min {
> > -    return min;
> > -  }
> > -  if n > max {
> > -    return max;
> > -  }
> > -  n
> 
> n is the input, in rust you dont have to return n; you can just leave it
> at n and it means the same
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/calebmer/accelerate/pull/2/files#r48104873.

svmnotn replied:
yup... not all rust statements need one (this mostly applies to one liners though)

svmnotn committed:
get tests passing again + redo some to use less static knowledge of how many motions the test suit has

svmnotn committed:
simplify macro use

calebmer committed:
Document accelerator

svmnotn committed:
added test files

svmnotn committed:
move motions into mod and split off template

This also marks the end of the read part of motions! :D

svmnotn committed:
prettify ls

svmnotn committed:
implement motions::discover :D

svmnotn committed:
add shift and fix goto help text

svmnotn committed:
fix rust types because ... rust

svmnotn committed:
begin implementing create

svmnotn committed:
implement create

svmnotn committed:
finish implementing create

svmnotn committed:
format

calebmer committed:
Add editor config

calebmer committed:
Rearrange functions

calebmer commented on src/motions/mod.rs:
```diff
@@ -0,0 +1,139 @@
+
+pub fn create(directory: String, name: String) {
+  let cookie = template::Template::get(&directory, &read_directory(&directory));
+  let motion_last = discover(&directory).pop().unwrap();
```
`Template::get` is called again in the discover function. Maybe make `cookie` a parameter?

svmnotn replied:
I wish, I still need to revise that code, but essentially because we have to call discover by itself I can't do that what I did think is that I could just pass in the result of it since I have it in the caller. Which is what I will do now

calebmer committed:
Add motions documentation

calebmer commented on src/motions/mod.rs:
```diff
@@ -0,0 +1,139 @@
+  while let Some(n) = motion_names.pop() {
+    match cookie.get_op(&n) {
+      template::Op::Add => motions_add.push(n),
+      template::Op::Sub => motions_sub.push(n),
```
Yeah, we really need to move operations to its own module. In order to combine the two, just do this:

``` rust
pub enum Operation {
  Add,
  Sub,
}

impl Operation {
  fn get(finish: isize, start: isize) -> Self {
    if finish < start {
      return Operation::Sub;
    }
    Operation::Add
  }

  fn get_direction(&self) -> isize {
    match self {
      Add => 1,
      Sub => -1,
    }
  }
}
```

Note the new `get_direction` function.

calebmer committed:
Add template documentation

calebmer committed:
Document the drivers

calebmer committed:
Remove XXX

Prefer a comment on the pull request.

calebmer commented on src/drivers/mod.rs:
```diff
@@ -0,0 +1,39 @@
+  fn execute(&self, motion: &String);
+}
+
+pub struct DefaultDriver {
```
This should be either in its own file or the test directory.

svmnotn replied:
Im using Default as a sort of when there is nothing else use this back up so that things can work without me implementing postgres, however I can move it to its own file.

calebmer commented on src/motions/mod.rs:
```diff
@@ -0,0 +1,139 @@
+    }
+  }
+
+  pub fn test(n: usize) -> Self {
```
This should be in the test directory.

svmnotn replied:
I'll try but I'm not sure if I can impl Motion twice

svmnotn replied:
Yes, yes I can.

calebmer commented on src/main.rs:
```diff
@@ -0,0 +1,78 @@
+  }
+}
+
+fn create(directory: String, name: String) { motions::create(directory, name); }
```
Could you expand this function into three lines? Generally that‘s good practice for readability and code coverage tests.

svmnotn replied:
I'll inline it into the match above since it is a single line anyways

svmnotn committed:
Merge pull request #1 from calebmer/docs/inline

Document accelerator

svmnotn committed:
move Default driver into its own module

svmnotn committed:
use new location of DefDriver in tests

svmnotn committed:
impl Display for Motion so that it can print all its info

svmnotn committed:
increase ls functionality given new printing options of Motion

svmnotn committed:
better errors and general code for create

svmnotn committed:
fix merge

svmnotn committed:
fix comment weirdness with " and '

svmnotn committed:
fix merger/rebase bug

svmnotn committed:
panic if length is more than max

svmnotn committed:
added cmd files to make sure that tests pass and that code is clean

svmnotn committed:
remove motions folder in test for less typing and add a 4.x.x folder to semantic to test folders

svmnotn committed:
implement reading from sub folders

svmnotn committed:
stop showing debug path

svmnotn committed:
impl gated subcommands so that deletions would have to be accepted

svmnotn committed:
fix strangeness of down working when m should be None

svmnotn committed:
derive all the things!

svmnotn committed:
use cargo fmt instead of directly calling rustfmt since I finally figured why it was being a pain with the write-mode.

Using the latest rustfmt (the one from github not crates.io) fixes this issue

svmnotn committed:
add .travis.yml and make target a pos arg

calebmer commented on build.cmd:
```diff
@@ -0,0 +1,4 @@
+cargo fmt -- --write-mode=overwrite
+cargo build
```
Is there a way to write this so it's interoperable across os? I usually use bash (`.sh`) scripts. Also I generally like to put the scripts in their own directory, so `scripts/build.cmd` and `scripts/clean.cmd`.

svmnotn replied:
I will be changing this to a .sh as soon as I make the switch to Arch

calebmer commented on src/drivers/postgres.rs:
```diff
@@ -0,0 +1,19 @@
+  fn new(target: String) -> Self { Driver { target: target } }
+}
+
+// TODO Implement
```
Yeah, this still needs to be done.

calebmer commented on src/main.rs:
```diff
@@ -0,0 +1,113 @@
+    for line in stdin.lock().lines() {
+      let s: &str = &line.unwrap();
+      match s {
+        "Yes" | "yes" | "Y" | "y" | "yup" => {
```
😊

calebmer commented:
Nice! What's left to do:
- Add the motions discovery tests.
- Add tests for new features.
- Implement postgres driver.
- Releases?

calebmer commented:
Also, is there anything I need to doc or fix docs for? If so just comment the areas.

svmnotn commented:
I would like you to go over the code and add doc for anything that does not have it, also some reorganizing would be nice.

svmnotn committed:
only use \n

svmnotn committed:
cache .cargo

calebmer commented:
Sorry for not responding for a while, school takes up too much time…

Did a code review, it's looking great! One thing I noticed, PostgreSQL support hasn't been done yet, but that shouldn't be too hard. I'll write docs, but it might be a bit (school makes it harder to have writing time & coding time).

---

I talked to someone the other day about database migrations and found out a couple of directions accelerate should probably take. There are two actionable things I derived, both of which have to do with better support for a git multi-developer environment, and both of which you should probably weigh in on. I should be able to help develop these things once we merge in what we have currently (it gets kinda crazy working on a pr to a pr :blush:). Here are those things:
1. Instead of doing `001` -> `002` for sorting motions, we should use a timestamp with seconds precision. This should give enough entropy that if two developers both open a PR with two new motions, their versions shouldn't collide.
2. State should probably not be represented as a number. Again let's use the example of two developers with two PRs with new motions—after they both merge in one of them will have a new motion in the middle of their motions list and accelerate will try to reapply some motions as the number and order of motions changed. We should implement a method to accommodate the use case when the number of motions and the order of the motions changes. Not exactly sure how we do this. Maybe a list of added motions we can diff against?
